### PR TITLE
Fix use after free of task in FuturesUnordered when dropped future panics

### PR DESCRIPTION
**Pull request by Imberflur**
Fixes https://github.com/rust-lang/futures-rs/issues/2863

### CODE REVIEW

File: futures-util/src/stream/futures_unordered/mod.rs [hunk 1]
```diff
@@ -276,8 +266,25 @@ impl<Fut> FuturesUnordered<Fut> {
...
+        //
+        // Use ManuallyDrop to transfer the reference count ownership before
+        // dropping the future so unwinding won't release the reference count.
+        let md_slot;
+        let task = if prev {
+            md_slot = mem::ManuallyDrop::new(task);
+            &*md_slot
+        } else {
+            &task
+        };
```
Comment (Imberflur) [inline futures-util/src/stream/futures_unordered/mod.rs:278]: This is the main fix

File: futures-util/src/stream/futures_unordered/mod.rs [hunk 2]
```diff
@@ -566,15 +573,36 @@ impl<Fut> FuturesUnordered<Fut> {
...
+            }
+        }
+        impl<Fut> Drop for DropGuard<'_, Fut> {
+            fn drop(&mut self) {
+                self.drop_futures();
+            }
         }
+        let mut guard = DropGuard(self);
+        guard.drop_futures();
+        mem::forget(guard); // no need to check head_all again
```
Comment (Imberflur) [inline futures-util/src/stream/futures_unordered/mod.rs:605]: This was added to make sure all futures are dropped if one panics. Although it looks like tasks have a check for this in their drop that aborts. At least with this, leaking and aborts may be avoided if only one future panics.

File: futures/tests/stream_futures_unordered.rs [hunk 3]
```diff
@@ -406,3 +406,26 @@ fn clear_in_loop() {
         }
     });
 }
+
+// https://github.com/rust-lang/futures-rs/issues/2863#issuecomment-2219441515
+#[test]
+#[should_panic]
+fn panic_on_drop_fut() {
```
Comment (Imberflur) [inline futures/tests/stream_futures_unordered.rs:413]: Minimal example taken from issue. This failed under miri before the changes to fix it.

**Review by Imberflur — COMMENTED** [review]
Notes for reviewers

File: futures-util/src/stream/futures_unordered/mod.rs [hunk 4]
```diff
@@ -566,15 +573,36 @@ impl<Fut> FuturesUnordered<Fut> {
...
-            let task = unsafe { self.unlink(head) };
-            self.release_task(task);
+        // Before the strong reference to the queue is dropped we need all
+        // futures to be dropped. See note at the bottom of this method.
+        //
+        // This ensures we drop all futures even in the case of one drop
+        // panicking and unwinding.
+        //
+        // If a second future panics, that will trigger an abort from panicking
+        // while panicking.
```
Comment (taiki-e) [inline futures-util/src/stream/futures_unordered/mod.rs:583]: If possible, I would prefer to leak the remaining elements after the first panic has occurred. That is the same behavior as in the standard library and crossbeam's queues/channels (https://github.com/rust-lang/rust/pull/108164#discussion_r1134233602).

**Review by taiki-e — CHANGES_REQUESTED** [review]
Thanks! The fix looks good to me.

File: futures-util/src/stream/futures_unordered/mod.rs (hunk 4 above)
Comment (Imberflur) [reply futures-util/src/stream/futures_unordered/mod.rs:583]: Hmm miri is complaining about the leak.

With
```sh
MIRIFLAGS="-Zmiri-strict-provenance -Zmiri-symbolic-alignment-check -Zmiri-disable-isolation" cargo +nightly miri nextest run --workspace --all-features -- panic_on_drop_fut
```

Comment (Imberflur) [reply futures-util/src/stream/futures_unordered/mod.rs]: I added a separate miri invocation for the test that ignores leaks

Comment (Imberflur) [reply futures-util/src/stream/futures_unordered/mod.rs]: Btw there might be some potential to make the miri CI faster using nextest

**Review by taiki-e — APPROVED** [review]
Thanks!

**Comment by taiki-e** [comment]
Hmm. I see AddressSanitizer also complains about the same leak.

Given that it would complicate CI, and how Vec and VecDeque behave, trying to avoid the risk of potential abort for bad future here may not actually be a worthwhile goal...

**Comment by Imberflur** [comment]
I could just ignore this test there? The main thing it tests for is caught by miri.

It feels kind of bad to change this due to testing issues....

TBH if there was a test for hitting the abort case that would probably have issues too since iirc there isn't a good way to test `should_abort` :(

**Review by taiki-e — APPROVED** [review]

**Pull request merged**